# Implementing Apriori Algorithm for Association Rule Mining

In [3]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [21]:
# Importing the Dataset. My data has not header and I specify that header=None
data = pd.read_csv("Market_Basket_Optimisation.csv", header=None)

In [22]:
#Print top n rows from our dataset
data.head()

0          1           2                 3             4   \
0         shrimp        nut       lemon    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey      lemon         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice      iced tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10        11     12     13             14      15  \
0  low fat yogurt  iced tea  honey  salad  mineral water  salmon   
1             NaN       NaN    NaN    NaN            NaN     NaN   
2             NaN       NaN    NaN    NaN            NaN     NaN   
3             NaN       NaN    NaN    NaN            NaN     NaN   
4             NaN       NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [23]:
#Check how many rows and columns we have in our dataset
data.shape

(7501, 20)

## Install apyori library

In [6]:
#We need to install apyori becasue scikit-learn library doesn't include APRIORI algorithm 
!pip install apyori

## Data Pre-Proccessing

The input data for Apriori should be the list, not the pandas dataframe. So we need to convert our dataframe into list which contains sublists. For this we need to create a loop which will go through all rows and all columns. 

At first we need to create an empty list. After creating an empty list we need to append the list with the elements in our dataset converted into string using loop. We have 7501 rows and 20 columns. So variable i should start from 0 and go to 7501. Then, for each row we need to look at 20 columns. Thats why we are using the second for cycle inside the loop which starts from 0 and goes to 20

In [24]:
#Let's create an empty list here
list_of_transactions = []
#Append the list
for i in range(0, data.shape[0]):
    list_of_transactions.append([str(data.values[i,j]) for j in range(0, data.shape[1])])


In [25]:
#Let's see the first element from our list of transactions. We should indicate 0 here because index in Pythn starts with 0
list_of_transactions[0]

['shrimp',
 'nut',
 'lemon',
 'vegetables mix',
 'green grapes',
 'whole weat flour',
 'yams',
 'cottage cheese',
 'energy drink',
 'tomato juice',
 'low fat yogurt',
 'iced tea',
 'honey',
 'salad',
 'mineral water',
 'salmon',
 'antioxydant juice',
 'frozen smoothie',
 'spinach',
 'olive oil']

## Training Apriori Algorithm

In [26]:
# Training apiori algorithm on our list_of_transactions
from apyori import apriori
rules = apriori(list_of_transactions, min_support = 0.004, min_confidence = 0.2, min_lift = 3, min_length = 2)
#So we will train apriori algorithm on our list_of_transactions and get the rules where items appear together minimum 0.004 times in total transaction (support) and there is minimum 20% chance that item will be added on the cart after purchasing base item (confidence), minimum lift is 3 and minimum number of items in rule is 2

In [27]:
# Create a list of rules and print the results
results = list(rules)

In [28]:
#Here is the first rule in list or results
results[0]

RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)])

Let's discuss the first rule -> {'chicken', 'light cream'} with support=0.0045, confidence=0.291 and lift=4.84. Please pay attention to that: items_base is {'light cream'} and items_add is {'chicken'}. This means that there is 29% chance (confidence) that user will buy chicken if he has already bought light cream. So left hand side is light cream and right hand side is chicken.

## Putting the results into a Pandas DataFrame

In [30]:
#In order to visualize our rules better we need to extract elements from our results list, convert it to pd.data frame and sort strong rules by lift value.
#Here is the code for this. We have extracted left hand side and right hand side items from our rules above, also their support, confidence and lift value
def inspect(results):
    lhs     =  [tuple(result [2] [0] [0]) [0] for result in results]
    rhs     =  [tuple(result [2] [0] [1]) [0] for result in results]
    supports = [result [1] for result in results]
    confidences = [result [2] [0] [2]   for result in results]
    lifts = [result [2] [0] [3]   for result in results]
    return list(zip(lhs,rhs,supports,confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results),columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'] )
resultsinDataFrame.head(5)

Left Hand Side Right Hand Side   Support  Confidence      Lift
0           light cream         chicken  0.004533    0.290598  4.843951
1  mushroom cream sauce        escalope  0.005733    0.300699  3.790833
2                 pasta        escalope  0.005866    0.372881  4.700812
3         herb & pepper     ground beef  0.015998    0.323450  3.291994
4          tomato sauce     ground beef  0.005333    0.377358  3.840659

In [31]:
#As we have our rules in pd.dataframe we can sort it by lift value using nlargest command. Here we are saying that we need top 6 rule by lift value
resultsinDataFrame.nlargest(n=6, columns='Lift')

Left Hand Side Right Hand Side   Support  Confidence      Lift
0     light cream         chicken  0.004533    0.290598  4.843951
7     light cream         chicken  0.004533    0.290598  4.843951
2           pasta        escalope  0.005866    0.372881  4.700812
12          pasta        escalope  0.005866    0.372881  4.700812
30          pasta             nan  0.005066    0.322034  4.515096
6           pasta          shrimp  0.005066    0.322034  4.506672

## ECLAT

In [ ]:
!pip install pyECLAT

In [33]:
# store the item sets as lists of strings in a list
transactions = [
    ['beer', 'wine', 'cheese'],
    ['beer', 'potato chips'],
    ['eggs', 'flower', 'butter', 'cheese'],
    ['eggs', 'flower', 'butter', 'beer', 'potato chips'],
    ['wine', 'cheese'],
    ['potato chips'],
    ['eggs', 'flower', 'butter', 'wine', 'cheese'],
    ['eggs', 'flower', 'butter', 'beer', 'potato chips'],
    ['wine', 'beer'],
    ['beer', 'potato chips'],
    ['butter', 'eggs'],
    ['beer', 'potato chips'],
    ['flower', 'eggs'],
    ['beer', 'potato chips'],
    ['eggs', 'flower', 'butter', 'wine', 'cheese'],
    ['beer', 'wine', 'potato chips', 'cheese'],
    ['wine', 'cheese'],
    ['beer', 'potato chips'],
    ['wine', 'cheese'],
    ['beer', 'potato chips']
]

In [34]:
import pandas as pd

# you simply convert the transaction list into a dataframe
data1 = pd.DataFrame(transactions)
data1

0             1             2       3             4
0           beer          wine        cheese    None          None
1           beer  potato chips          None    None          None
2           eggs        flower        butter  cheese          None
3           eggs        flower        butter    beer  potato chips
4           wine        cheese          None    None          None
5   potato chips          None          None    None          None
6           eggs        flower        butter    wine        cheese
7           eggs        flower        butter    beer  potato chips
8           wine          beer          None    None          None
9           beer  potato chips          None    None          None
10        butter          eggs          None    None          None
11          beer  potato chips          None    None          None
12        flower          eggs          None    None          None
13          beer  potato chips          None    None          None
14          eggs        flower        butter    wine        cheese
15          beer          wine  potato chips  cheese          None
16          wine        cheese          None    None          None
17          beer  potato chips          None    None          None
18          wine        cheese          None    None          None
19          beer  potato chips          None    None          None

In [39]:
# we are looking for itemSETS
# we do not want to have any individual products returned
min_n_products = 2

# we want to set min support to 7
# but we have to express it as a percentage
min_support = 7/len(transactions)

# we have no limit on the size of association rules
# so we set it to the longest transaction
max_length = max([len(x) for x in transactions])

In [40]:
max_length

5

In [41]:
from pyECLAT import ECLAT

# create an instance of eclat
my_eclat = ECLAT(data=data1, verbose=True)

# fit the algorithm
rule_indices, rule_supports = my_eclat.fit(min_support=min_support,
                                           min_combination=min_n_products,
                                           max_combination=max_length)

100%|██████████| 8/8 [00:00<00:00, 1328.36it/s]


Combination 2 by 2


10it [00:00, 127.04it/s]


Combination 3 by 3


10it [00:00, 119.37it/s]


Combination 4 by 4


5it [00:00, 41.09it/s]


Combination 5 by 5


1it [00:00, 77.41it/s]


In [42]:
print(rule_supports)


{'wine & cheese': 0.35, 'beer & potato chips': 0.45}


The interpretation of this is that within the transactions of our night store, there are two product combinations that are relatively strong. People often buy Wine and Cheese together. People also often buy Potato Chips and Beer together. Clearly, it could be a good idea to put those products together so that people can easily get to both of them. Or maybe the shop owner could think about packaging the products in an attractive offer to boost sales of those products even more.